docs(root): add guidance and code examples for popover menu component (#391)
mi6/ic-design-system: src/content/structured/components/popover-menu/guidance.mdx

      </IcPopoverMenu>
      <IcPopoverMenu submenuId="actions">
        <IcMenuItem label="Edit" />
        <IcMenuItem label="Find" submenuTriggerFor="find" />
        <IcMenuItem label="Delete" variant="destructive" />
      </IcPopoverMenu>
      <IcPopoverMenu submenuId="find">
        <IcMenuItem
          label="Search the web"
          description="This will search the web to find the thing you are looking for."
        />
        <IcMenuItem label="Find icons">
          <svg
            slot="icon"
            xmlns="http://www.w3.org/2000/svg"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
          >
            <path d="M0 0h24v24H0V0z" fill="none" />
            <path d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z" />
          </svg>
        </IcMenuItem>
        <IcMenuItem label="Show found results" variant="toggle" />
      </IcPopoverMenu>
    </>
  );
};

<ComponentPreview
  style={{
    display: "flex",
    justifyContent: "flex-start",
  }}
>
  <IntroPopover />

> </ComponentPreview>

Reviewer: In addition, could we make the intro example simpler? It's a bit confusing seeing all the different variants when first viewing the component. Perhaps copy an example from the screenshot below. Then when consumers want to see more complex examples, they can view them in code > variants.

> </ComponentPreview>

Author: The button should centre without moving now if you get the components from the popover-fixes branch

## When to use

Use a popover menu to place multiple actions behind a single trigger. This is especially useful when there is not enough space to include all actions on the base page. Popover menus are commonly used for ‘More actions’ and often triggered using an icon button.

<DoDontCaution
  imageSrc={popoverFig1}
  imageAlt="A popover menu with edit and view options using a mix of menu items, menu groups and menu toggle items."
  state="good"
  caption="Use popover menus to contain multiple actions."
/>

## When not to use

Don’t use a popover menu if there are only a small number of actions. Consider if the actions can be better placed on the page itself.

> Avoid placing important actions in popover menus as these may be missed. Instead use popover menus to house secondary or less important actions and place your most important actions on the page itself.

Reviewer 2: ```suggestion
Avoid placing important actions in popover menus as these may be missed. Instead, use popover menus to house secondary or less important actions and place your most important actions on the page itself.
```

  imageSrc={popoverFig3}
  imageAlt="Two examples of popover menus. The first displays all menu items within the menu container. The second example shows the same menu items, but with some hidden due to a smaller menu container. A scroll bar allows the menu to be scrolled."
  state="good"
  caption="Set a max-height to constrain the height of the menu and cause its contents to scroll."
/>

### Layout and placement

Popover menus appear adjacent to their triggering element. By default, they appear below and aligned to the left of the trigger element. However, if no space is available, the popover menu will appear above or aligned to the right of the trigger element.

<DoDontCaution
  imageSrc={popoverFig4}
  imageAlt="Two examples of popover menus. The first example shows the menu aligned to the left of the screen. The second shows the menu aligned to the right."
  state="good"
  caption="Popover menus position themselves based on the available space."
/>

## Content

Popover menus contain menu items, menu toggle items, menu groups and menu sections.

### Labels

Follow the [content style](/styles/content-style) guidelines when writing labels and keep them concise but specific.

> Always specify an accessible label for the menu items. The accessible label can include more detail than is described in the visible label, but make sure that the visible label’s text is included within the accessible label.

Reviewer 2: ```suggestion
Always specify an accessible label for the menu items. The accessible label can include more detail than is described in the visible label.
```

> For example, a menu item may have a visible label which reads ‘Edit’. The accessible label may read ‘Edit record’. This includes more information to aid assistive technology users, but still includes the full visible label.

Reviewer 2: ```suggestion
For example, a menu item may have a visible label which reads ‘Edit’. The accessible label may read ‘Edit record’. This includes more information to aid assistive technology users.
```

### Icons

Use an icon on a menu item to provide additional meaning or to aid in recognition. Menu items, menu toggle items and parent menu items can display an icon.

Make sure that the icon is understandable and relevant to the action.

Always provide alternative descriptions for icons if they aren't decorative. Only use icons if they are necessary.

### Menu items

> Use menu items for single actions that do not have a selection status.

Reviewer 2: ```suggestion
Use the "default" variant of menu items for single actions that do not have a selection status.
```

Add an additional description to a menu item to provide additional context. Keep the description concise and consider if the action is too complicated if too much information is required to describe the action.

If a keyboard shortcut is available for a menu item’s action, then display it using the keyboard shortcut label.

> Use a destructive menu item to highlight a destructive action. Position destructive actions at the bottom of a popover menu.

Reviewer 2: ```suggestion
Use the "destructive" variant of a menu item to highlight a destructive action. Position destructive actions at the bottom of a popover menu.
```

Specify a menu item as a parent if it has children that appear in a sub-menu. When clicked, the sub-menu will slide into view.

<DoDontCaution
  imageSrc={popoverFig5}
  imageAlt="A popover menu showing a set of edit options that include icons and keyboard shortcuts alongside their labels."
  state="good"
  caption="Use menu items for single actions. Add an icon or a description to provide more context. Highlight keyboard shortcuts using the keyboard shortcut label."
/>

### Menu toggle items

> Use menu toggle items for actions that provide an ability to toggle something on or off. Menu toggle items will display a checkmark when toggled on and hide it when toggled off.

Reviewer 2: ```suggestion
Use the "toggle" variant of menu items for actions that provide an ability to toggle something on or off. Menu toggle items will display a checkmark when toggled on and hide it when toggled off.
```

> Like menu items, provide an icon or a description to add additional context if required.

Reviewer 2: ```suggestion
Like default menu items, provide an icon or a description to add additional context if required.
```

<DoDontCaution
  imageSrc={popoverFig6}
  imageAlt="A popover menu containing two menu toggle items. Their labels read ‘Show/Hide’ and ‘Lock/Unlock’ and display a tick indicating that they are toggled on."
  state="good"
  caption="Use menu toggle items to show a selection state alongside the action."
/>

### Menu groups

Use menu groups to organise menu items into sections within the popover menu. Provide a label for the menu group to describe its contents. Alternatively, leave the menu group label blank if a description is unnecessary.

<DoDontCaution
  imageSrc={popoverFig7}
  imageAlt=" A popover menu with a menu group for ‘Edit options’, a menu group for ‘View options’, an untitled menu group containing unrelated actions, and a final untitled menu group containing destructive actions."
  state="good"
  caption="Use menu groups to create sections within a popover menu."
/>

### Parent menu items

> Use a parent menu item to open sub-menus including more menu items. Use this to contain similar actions or options relating to the menu group name.

Reviewer 2: ```suggestion
Use a parent menu item to open sub-menus, including more menu items. Use this to contain similar actions or options relating to the menu group name.
```

<video
  width="100%"
  muted
  controls
  loop

>   title="An animation showing a popover menu with a parent menu item titled ‘Zoom’. A submenu containing menu items for different zoom levels is opened after the ‘zoom’ parent menu item is clicked. After the back button is clicked in the submenu the main menu is displayed again."

Reviewer 2: ```suggestion
  title="An animation showing a popover menu with a parent menu item titled ‘Zoom’. A sub-menu containing menu items for different zoom levels is opened after the ‘zoom’ parent menu item is clicked. After the back button is clicked in the sub-menu the main menu is displayed again."
```

>
  <source src={popoverVid2} type="video/mp4" />
</video>

### Sub-menus

Use a sub-menu to contain a parent menu item’s children items. Use sub-menus for multiple related actions that are opened from the main menu.

> Make sure the sub-menu has the same label as the parent menu item. Clicking the back button in a sub-menu will return the user to the parent menu item in the main menu.

Reviewer 2: ```suggestion
The sub-menu displays the same label as the parent menu item. Clicking the back button in a sub-menu will return the user to the parent menu item in the main menu.
```

Use menu items, toggle menu items, menu groups or parent menu items in a sub-menu but avoid nesting too many levels of sub-menus as this can be confusing to navigate.

<video
  width="100%"
  muted
  controls
  loop
  title="An animation showing a popover menu that contains four parent menu items each with nested parent menu items in their sub-menus. The user clicks ‘help options’ to go into the sub-menu, then clicks ‘information’ to go into a second sub-menu, then clicks ‘tutorials’ to go into a third sub-menu."
>
  <source src={popoverVid3} type="video/mp4" />
</video>

## Accessibility considerations

> When specifying the trigger element for a popover menu, make use of ‘aria-expanded’ and ‘aria-collapsed’ so that a screen reader announces the appearance of the popover menu.

Reviewer: Is this not already built into the component so the consumer is unable to not use the aria attributes?

Author: This isn't built in functionality, since we won't be sure what component consumers are using to display the popover, it seemed the best way to make sure they add aria-expanded and aria-collapsed was to add it to the guidance.
